### ISSUE TITLE
Distributed sampling fails on heterogeneous graphs when the number of edge type is larger than the fanout.

### ISSUE THREAD

**Issue by zheng-da** [dmlc/dgl#4003]
## 🐛 Bug

The error happens when we perform distributed training on MAG-LSC graph and the fanout for neighbor sampling is 3. That is, DGL should sample 3 neighbors for each relation type.

```
Traceback (most recent call last):
  File "m5gnn_nc_huggingface.py", line 25, in <module>
    main(args)
  File "m5gnn_nc_huggingface.py", line 13, in main
    trainer = M5gnnNodePredictTrainer(config, m5_models)
  File "/m5-gnn/python/m5gnn/trainer/np_trainer.py", line 50, in __init__
    self.init_dist_context(config.ip_config,
  File "/m5-gnn/python/m5gnn/trainer/m5gnn_trainer.py", line 20, in init_dist_context
    dgl.distributed.initialize(ip_config)
  File "/usr/local/lib/python3.8/dist-packages/dgl/distributed/dist_context.py", line 235, in initialize
    serv.start()
  File "/usr/local/lib/python3.8/dist-packages/dgl/distributed/dist_graph.py", line 368, in start
    start_server(server_id=self.server_id,
  File "/usr/local/lib/python3.8/dist-packages/dgl/distributed/rpc_server.py", line 92, in start_server
    res = req.process_request(server_state)
  File "/usr/local/lib/python3.8/dist-packages/dgl/distributed/graph_services.py", line 189, in process_request
    global_src, global_dst, global_eids = _sample_etype_neighbors(local_g,
  File "/usr/local/lib/python3.8/dist-packages/dgl/distributed/graph_services.py", line 87, in _sample_etype_neighbors
    sampled_graph = local_sample_etype_neighbors(
  File "/usr/local/lib/python3.8/dist-packages/dgl/sampling/neighbor.py", line 123, in sample_etype_neighbors
    subgidx = _CAPI_DGLSampleNeighborsEType(g._graph, nodes, etypes, fanout,
  File "/usr/local/lib/python3.8/dist-packages/dgl/_ffi/_ctypes/function.py", line 188, in __call__
    check_call(_LIB.DGLFuncCall(
  File "/usr/local/lib/python3.8/dist-packages/dgl/_ffi/base.py", line 64, in check_call
    raise DGLError(py_str(_LIB.DGLGetLastError()))
dgl._ffi.base.DGLError: [00:04:34] /root/dgl/src/array/cpu/./rowwise_pick.h:273: Check failed: et[et_idx[len - 1]] < num_etypes: etype values exceed the number of fanouts
```

## Environment

 - DGL Version (e.g., 1.0):
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3):
 - OS (e.g., Linux):
 - How you installed DGL (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version (if applicable):
 - GPU models and configuration (e.g. V100):
 - Any other relevant information:

**Comment by BarclayII** [comment]
The error is saying that the length of fanout list (same as `len(g.etypes)`) is smaller than the edge type IDs on the homogenized heterogeneous graph.  Could you check if the edge type ID values are less than `len(g.etypes)` (which it should)?

**Comment by jermainewang** [comment]
> The error is saying that the length of fanout list (same as `len(g.etypes)`) is smaller than the edge type IDs on the homogenized heterogeneous graph.

The error message doesn't sound right though.

**Comment by jermainewang** [comment]
@BarclayII The issue seems to happen when calling distributed `sample_etype_neighbors` on a heterograph but only providing one integer as fanout (instead of a dict) to sample same number of neighbors for all relations.

**Comment by BarclayII** [comment]
@jermainewang I don't think so since distributed `sample_etype_neighbors` will expand the `int` to a tensor with the same number of elements as the number of edge types:
https://github.com/dmlc/dgl/blob/4c14781427c28be8d6e4b9b99a1b17e11b51460a/python/dgl/distributed/graph_services.py#L483-L484
@zheng-da How did you call `sample_etype_neighbors`?  Could you print the type of the fanout?  Maybe it's not a Python integer?

**Comment by BarclayII** [comment]
Recap on what I have found so far:
@classicsong gave me a partitioned MAG240M example and I reproduced the bug.  However, I'm confident that the problem is not in `sample_etype_neighbors`.
So I added these statements in the RGCN training script in the `main` function, before `run`: (https://github.com/BarclayII/dgl/blob/35a55f460963b005c1f94efd027bfa1e7a55d72d/python/dgl/sampling/neighbor.py#L118-L125)
```python
    local_g = g.local_partition
    logger.info('master %s' % local_g)
    for k, v in local_g.ndata.items():
        logger.info('ndata[%s] = (min:%s, max:%s)' % (k, v.min(), v.max()))
    for k, v in local_g.edata.items():
        logger.info('edata[%s] = (min:%s, max:%s)' % (k, v.min(), v.max()))

    return
```
Using PyTorch 1.11.0 + CUDA 10.2 (CUDA probably doesn't matter), and compiling DGL with Cython and
```
cmake -DUSE_CUDA=ON -DUSE_OPENMP=ON -DBUILD_TORCH=ON -DCUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda-10.2 ..
```
and running the script with (note that `num_samplers` is 0)
```
python3 /mnt/efs/gq/dgl/tools/launch.py --workspace /mnt/efs/gq/dgl/examples/pytorch/rgcn/experimental/ --num_trainers 1 --num_servers 1 --num_samplers 0 --part_config /mnt/efs/gq/distdgl/mag-lsc.json --ip_config ip_config.txt "/mnt/efs/gq/conda-env-gq/bin/python3 entity_classify_dist.py --graph-name mag-lsc --dataset mag-lsc --fanout='25,25' --batch-size 1024  --n-hidden 64 --lr 0.01 --eval-batch-size 1024  --low-mem --dropout 0.5 --use-self-loop --n-bases 2 --n-epochs 3 --layer-norm --ip-config ip_config.txt  --sparse-embedding --sparse-lr 0.06 --num_gpus 1 --dgl-sparse"
```
The code crashed with `SIGBUS` right at printing `edata`'s maximum and minimum value.
So the problem is likely that the local partition's `edata` got corrupted at some point, even well before training starts.

**Comment by BarclayII** [comment]
Confirmed that this issue will only happen when the stored edge type IDs as well as the `inner_edge` array are not int32 arrays.  Labeling as won't fix and will track in #4044.
